File: ts-reference/tests/example-program-gen/exp/errors/custom.ts
export type CustomError = SomeError | OtherError

export class SomeError extends Error {
  readonly code = 6000
  readonly name = "SomeError"
  readonly msg = "Example error."

  constructor() {
    super("6000: Example error.")
  }
}

export class OtherError extends Error {
  readonly code = 6001
  readonly name = "OtherError"
  readonly msg = "Another error."

  constructor() {
    super("6001: Another error.")
  }
}

export function fromCode(code: number): CustomError | null {
  switch (code) {
    case 6000:
      return new SomeError()
    case 6001:
      return new OtherError()
  }

  return null
}

File: tests/client_gen/example_program_gen/errors/custom.py
import typing
from anchorpy.error import ProgramError


class SomeError(ProgramError):
    def __init__(self) -> None:
        super().__init__(6000, "Example error.")

    code = 6000
    name = "SomeError"
    msg = "Example error."


class OtherError(ProgramError):
    def __init__(self) -> None:
        super().__init__(6001, "Another error.")

    code = 6001
    name = "OtherError"
    msg = "Another error."


CustomError = typing.Union[SomeError, OtherError]
CUSTOM_ERROR_MAP: dict[int, CustomError] = {
    6000: SomeError(),
    6001: OtherError(),
}


def from_code(code: int) -> typing.Optional[CustomError]:
    maybe_err = CUSTOM_ERROR_MAP.get(code)
    if maybe_err is None:
        return None
    return maybe_err
